src/ripple/app/tx/impl/DepositPreauth.cpp

```diff
@@ -0,0 +1,195 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/DepositPreauth.h>
+#include <ripple/basics/Log.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/st.h>
+#include <ripple/ledger/View.h>
+
+namespace ripple {
+
+TER
+DepositPreauth::preflight (PreflightContext const& ctx)
+{
+    if (! ctx.rules.enabled (featureDepositPreauth))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    std::uint32_t const uTxFlags = tx.getFlags ();
+
+    if (uTxFlags & tfUniversalMask)
+    {
+        JLOG(j.trace()) <<
+            "Malformed transaction: Invalid flags set.";
+        return temINVALID_FLAG;
+    }
+
+    auto const auth = ctx.tx[~sfAuthorize];
+    auto const unauth = ctx.tx[~sfUnauthorize];
+    if ((auth.is_initialized()) == unauth.is_initialized())
+    {
+        // Either both fields are present or neither field is present.  In
+        // either case the transaction is malformed.
+        JLOG(j.trace()) <<
+            "Malformed transaction: "
+            "Invalid Authorize and Unauthorize field combination.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    AccountID const target {auth.is_initialized() ? *auth : *unauth};
+    if (target == beast::zero)
+    {
+        JLOG(j.trace()) <<
+            "Malformed transaction: Authorized or Unauthorized field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    // An account may not preauthorize itself.
+    if (auth.is_initialized() && (target == ctx.tx[sfAccount]))
```

Comment:
No need for `is_initialized` here. We test optionals in if statements all the time.